### PR TITLE
Add hour view mode

### PR DESCRIPTION
**Pull request by anto-christo**
This pull request is in with reference to issue #52.
I have added the hour view mode.
`gantt_chart.change_view_mode('Hour')`

The date-time format should be as  (YYYY:MM:DD HH:MM:SS) `Eg:2018-03-24 00:00:00`

This is my first ever pull request. In case of any issues with it, please do let me know.

### CODE REVIEW

**Comment by juanarbol** [comment]
Please merge this! It's awesome

**Comment by netchampfaris** [comment]
@anto-christo This PR needs to be updated.

File: dist/frappe-gantt.js [hunk 1]
```diff
@@ -383,9 +383,20 @@ class Bar {
         this.x = this.compute_x();
         this.y = this.compute_y();
         this.corner_radius = this.gantt.options.bar_corner_radius;
-        this.duration =
-            (date_utils.diff(this.task._end, this.task._start, 'hour') + 24) /
+        console.log(this.gantt.options.view_mode);
```
Comment (jamieday) [inline dist/frappe-gantt.js]: is this a debug log left by mistake?

File: dist/frappe-gantt.js [hunk 2]
```diff
@@ -383,9 +383,20 @@ class Bar {
...
+            this.duration =
+            (date_utils.diff(this.task._end, this.task._start, 'hour')) /
+            this.gantt.options.step;
+        }
+
+        else{
+            this.duration =
+            (date_utils.diff(this.task._end, this.task._start, 'hour')+24) /
             this.gantt.options.step;
+        }
```
Comment (jamieday) [inline dist/frappe-gantt.js]: spacing here is inconsistent with the style of the rest of the code

File: dist/frappe-gantt.js [hunk 3]
```diff
@@ -383,9 +383,20 @@ class Bar {
...
+
+        if(this.gantt.options.view_mode==="Hour"){
+            this.duration =
+            (date_utils.diff(this.task._end, this.task._start, 'hour')) /
+            this.gantt.options.step;
+        }
+
+        else{
+            this.duration =
+            (date_utils.diff(this.task._end, this.task._start, 'hour')+24) /
```
Comment (jamieday) [inline dist/frappe-gantt.js]: spacing here is inconsistent with the style of the rest of the code

**Comment by anto-christo** [comment]
@jamieday , thanks for the review. Will take care of it when I update the PR.

File: dist/frappe-gantt.js (hunk 3 above)
Comment (jamieday) [reply dist/frappe-gantt.js]: try `npm run prettier`

**Comment by dikar** [comment]
who resolved this?

**Comment by Anurag810** [comment]
@anto-christo Please resolve the conflicts and make the changes requested.

**Comment by matthenning** [comment]
Any news @Anurag810 ?

**Comment by anto-christo** [comment]
Hi, I resumed working on this PR. However, I need to clear somethings first.

```
start: '2018-10-01',
end: '2018-10-01',
name: 'Redesign website',
id: "Task 0",
progress: 20
``` 

If this is how my task is like and when I view this in `Quarter Day` mode, it is displayed as below.

quarter-day

Is it right that the day is starting from '12' and not from '00'? I need to know this since I feel it will affect how `Hour View` will work as well. Thank you.

@Anurag810 @matthenning

**Comment by matthenning** [comment]
Days should start at 00 imho

**Comment by jcjobin** [comment]
> Hi, I resumed working on this PR. However, I need to clear somethings first.
> 
> ```
> start: '2018-10-01',
> end: '2018-10-01',
> name: 'Redesign website',
> id: "Task 0",
> progress: 20
> ```
> 
> If this is how my task is like and when I view this in `Quarter Day` mode, it is displayed as below.
> 
> quarter-day
> 
> Is it right that the day is starting from '12' and not from '00'? I need to know this since I feel it will affect how `Hour View` will work as well. Thank you.
> 
> @Anurag810 @matthenning

Just to add to this quickly, I ran into the same problem a while ago, if I remember correctly, it came from mis-calculating the x-positions. So within the get_date_info method on the Gantt class, I changed:

```
const x_pos = {
            'Quarter Day_lower': this.options.column_width * 4 / 2,
            'Quarter Day_upper': 0,
            'Half Day_lower': this.options.column_width * 2 / 2,
            'Half Day_upper': 0,
            Day_lower: this.options.column_width / 2,
            Day_upper: this.options.column_width * 30 / 2,
            Week_lower: 0,
            Week_upper: this.options.column_width * 4 / 2,
            Month_lower: this.options.column_width / 2,
            Month_upper: this.options.column_width * 12 / 2,
            Year_lower: this.options.column_width / 2,
            Year_upper: this.options.column_width * 30 / 2
        };
```

to:
```
const x_pos = {
            'Quarter Day_lower': 0,
            'Quarter Day_upper': this.options.column_width * 4 / 2,
            'Half Day_lower': 0,
            'Half Day_upper': this.options.column_width * 2 / 2,
            Day_lower: this.options.column_width / 2,
            Day_upper: this.options.column_width * 30 / 2,
            Week_lower: 0,
            Week_upper: this.options.column_width * 4 / 2,
            Month_lower: this.options.column_width / 2,
            Month_upper: this.options.column_width * 12 / 2,
            Year_lower: this.options.column_width / 2,
            Year_upper: this.options.column_width * 30 / 2
        };
```

Hope this helps!

**Comment by anto-christo** [comment]
@jcjobin 
Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.

Would you mind sending this fix as a PR? If merged, it would be helpful for others as well. 
Thank you.

**Comment by jcjobin** [comment]
> @jcjobin
> Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.
> 
> Would you mind sending this fix as a PR? If merged, it would be helpful for others as well.
> Thank you.

I'm not sure if I can create a PR, the button is disabled. (Doesn't help that I've never made one before either)

**Comment by jcjobin** [comment]
> @jcjobin
> Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.
> 
> Would you mind sending this fix as a PR? If merged, it would be helpful for others as well.
> Thank you.

Should I add the pull request to the refactor branch?

**Comment by anto-christo** [comment]
@jcjobin 
Would you mind if I use your solution for this PR? I can make the changes for Quarter Day and Half Day modes along with this PR since it just a couple of lines of change.

**Comment by jcjobin** [comment]
> @jcjobin
> Would you mind if I use your solution for this PR? I can make the changes for Quarter Day and Half Day modes along with this PR since it just a couple of lines of change.

Totally okay with me, I am not looking for street cred at this point ha.

**Comment by anto-christo** [comment]
> > @jcjobin
> > Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.
> > Would you mind sending this fix as a PR? If merged, it would be helpful for others as well.
> > Thank you.
> 
> Should I add the pull request to the refactor branch?

@jcjobin 
I am not sure on which branch the PR needs to be directed. I think it is ok if it is the master branch. In case you are sending the PR, let me know, then I need not incorporate these changes in my PR.
Thank you.

**Comment by jcjobin** [comment]
> > > @jcjobin
> > > Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.
> > > Would you mind sending this fix as a PR? If merged, it would be helpful for others as well.
> > > Thank you.
> > 
> > 
> > Should I add the pull request to the refactor branch?
> 
> @jcjobin
> I am not sure on which branch the PR needs to be directed. I think it is ok if it is the master branch. In case you are sending the PR, let me know, then I need not incorporate these changes in my PR.
> Thank you.

I won't be making a PR, so you can go for it. Sorry for the confusion. Personally, I have made a lot of changes locally by just extending the classes.

**Comment by anto-christo** [comment]
> > > > @jcjobin
> > > > Your solution fixes this perfectly. With similar changes, the tasks in the `Hour View` mode will also be aligned as expected.
> > > > Would you mind sending this fix as a PR? If merged, it would be helpful for others as well.
> > > > Thank you.
> > > 
> > > 
> > > Should I add the pull request to the refactor branch?
> > 
> > 
> > @jcjobin
> > I am not sure on which branch the PR needs to be directed. I think it is ok if it is the master branch. In case you are sending the PR, let me know, then I need not incorporate these changes in my PR.
> > Thank you.
> 
> I won't be making a PR, so you can go for it. Sorry for the confusion. Personally, I have made a lot of changes locally by just extending the classes.

Okay 👍 
I'll include it in my PR

**Comment by anto-christo** [comment]
@Anurag810 
I have updated the PR but I doubt whether I have done it correctly. It is showing all those changes after this PR was made as a part of this PR instead of showing only the changes that I made. Should I open a new PR altogether? Will it help?

**Comment by netchampfaris** [comment]
@anto-christo Hey, either rebase this with latest master or open a new PR. Let's get this in!

**Comment by anto-christo** [comment]
This PR has been moved to #142

**Pull request closed without merging**